### ISSUE TITLE
Mach exceptions in subprocesses cause myriad issues with the Sentry crash monitor

### ISSUE THREAD

**Issue by vorporeal** [getsentry/sentry-cocoa#2695]
### Platform

macOS

### Installed

Carthage

### Version

8.1.0

### Steps to Reproduce

1. Build a small binary that segfaults.  Can be done by compiling the following C code with `clang`:
```c
int main() {
  *(int*)0 = 0;
}
```
2. Create a small macOS binary that installs Sentry, executes the segfault binary as a subprocess, then crashes (e.g.: using `[SentrySDK crash]`).
3. Run the test binary.

### Expected Result

One crash report is generated and appears in the Sentry web UI - for the crash that occurred in the main application.  (I could imagine "report crashes from subprocesses" being configurable, if you don't want to unconditionally change this behavior for all users.)

### Actual Result

One crash report is generated - **the segfault in the subprocess**.  This crash report includes a stack trace from the _host_ process, wherever it happened to be in its execution at the time.  Notably, the subsequent crash in the host process itself is not captured, due to the fact that the exception handler doesn't wait on additional messages over the exception port after receiving one.

Here is an example Sentry report from a segfault in a subprocess: https://warpdotdev.sentry.io/issues/3938977232/?project=5660375&query=is:unresolved

I have a variety of concerns here:
* We receive a bogus crash report - a crash that did not occur in our application.
* The crash report we do receive claims to be from within our application, and includes a stack trace that never makes any sense.
* Some small amount of information _is_ provided from the actual crashing application - the subprocess - e.g.: memory addresses.  I don't think there would be any sensitive data in there, but in cases where the subprocess is an end-user application, we don't want to accidentally collect any information about it.  (We're building a terminal, and so can have arbitrary processes as descendants, and we never want to collect information about their internals.)

### Are you willing to submit a PR?

I have a proof-of-concept fix in this PR: https://github.com/getsentry/sentry-cocoa/pull/2691/files.  I manually tested this and confirmed that it produces the expected behavior, but I expect the Sentry team might want to make additional changes/add tests, and I'm not sure I have enough domain knowledge to do that additional work.

**Comment by vorporeal** [comment]
Also of note - I investigated a few different ways to work around this as a user, and there's no wonderful solution.

* I tried using `pthread_atfork` to add an at-fork handler to reset the task-level mach exception ports (to basically unhook Sentry), but the Rust runtime (our application is primarily Rust with some Obj-C bindings) uses `posix_spawn` where possible instead of fork/exec, and `posix_spawn` doesn't invoke at-fork handlers.
* Adding post-fork, pre-exec logic at every place we want to execute a subprocess is infeasible; there's no good way to prevent developers from using standard library subprocess execution functions as opposed to any wrapper that we would define.

While we probably could install our own mach exception handler after Sentry installs its own, and have it own this logic (delegating to Sentry for exceptions from the current process and ignoring exceptions from other processes), to me this feels like an issue Sentry should probably be addressing for all users.

**Comment by brustolin** [comment]
Thanks @vorporeal for you input and also the PR. We're going to look into it.

**Comment by brianmichel** [comment]
This is still a problem, no?

**Comment by philipphofmann** [comment]
> This is still a problem, no?

Yes, we didn't merge the PR https://github.com/getsentry/sentry-cocoa/pull/2691 yet.

**Comment by brianmichel** [comment]
> > This is still a problem, no?
> 
> Yes, we didn't merge the PR #2691 yet.

Sounds good, thanks @philipphofmann, when that change lands what is your recommendation for dealing with sub process crash reporting in general with Sentry?

**Comment by philipphofmann** [comment]
@brianmichel, I guess it makes sense to start the SDK again for the subprocess, or we add an option to make it configurable wether you would like to get crashes from subprocesses or not. It depends on the outcome of #2691. What would be your expected behavior @brianmichel?

**Comment by armcknight** [comment]
We closed #2691 for now until we can get a reproducible case and test the outcomes, so we can make a reasoned decision about the tradeoffs of the approach.

**Comment by dnguyen032123** [comment]
Hi, is there a way to make sentry-cocoa not report crashes of the sub-processes' at all? To give some context, my macOS app is using Chromium Embedded Framework (CEF), and several CEF sub-processes are spawned from a separate executable, not from the main process/app's executable. I'm thinking about using sentry-native for those CEF sub-processes and keep using sentry-cocoa for the main process.

**Comment by philipphofmann** [comment]
> Hi, is there a way to make sentry-cocoa not report crashes of the sub-processes' at all?

Not officially at the moment. You could use `beforeSend` to drop crashes from your sub-processes if you can identify them reliably, @dnguyen032123.

**Comment by dnguyen032123** [comment]
@philipphofmann Ok, thanks for the idea! The reason I asked was because in my case, the main process continues without being affected by the subprocesses' crashes. After that, if the main process crashes, we'll lose its crash report, not knowing that it crashed (because the crash report that'll be sent to the server at next launch is the subprocess, not the main process). It's not a big deal though. I'm just sharing my use case.

**Comment by kahest** [comment]
closing this as stale and potentially outdated
